modape/modis/download.py

```diff
@@ -172,45 +178,74 @@ def _parse_response(query: List[dict]) -> dict:
-                   url: str,
-                   destination: Path,
-                   overwrite: bool,
-                   ) -> Tuple[str, Union[None, Exception]]:
+    def _parse_hdfxml(response):
+        result = {}
+        tree = ElementTree.fromstring(response.content)
+        for entry in tree.iter(tag='GranuleURMetaData'):
+            for datafile in entry.iter(tag="DataFiles"):
+                for datafilecont in datafile.iter(tag="DataFileContainer"):
+                    for content in datafilecont:
+                        if content.tag in ["Checksum", "FileSize"]:
+                            result.update({content.tag: int(content.text)})
+        return result
+
+
+    def _fetch(self,
+               session: SessionWithHeaderRedirection,
+               url: str,
+               destination: Path,
+               overwrite: bool,
+               check: bool,
+               ) -> Tuple[str, Union[None, Exception]]:
         """Helper function to fetch HDF files
 
         Args:
             session (SessionWithHeaderRedirection): requests session to fetch file.
             url (str): URL for file.
             destination (Path): Target directory.
             overwrite (bool): Overwrite existing.
+            check (bool): Check file size and checksum.
 
         Returns:
             Tuple[str, Union[None, Exception]]: Returns tuple with
                 either (filename, None) for success and (URL, Exception) for error.
 
         """
-
-        filename = destination.joinpath(url.split("/")[-1])
+        file_id = url.split("/")[-1]
+        filename = destination.joinpath(file_id)
 
         if not exists(filename) or overwrite:
 
+            filename_temp = filename.with_suffix(".modapedl")
+
             try:
 
                 with session.get(url, stream=True, allow_redirects=True) as response:
                     response.raise_for_status()
 
-                    with open(filename, "wb") as openfile:
+                    with open(filename_temp, "wb") as openfile:
                         shutil.copyfileobj(response.raw, openfile, length=16*1024*1024)#
 
-                assert filename.exists(), "File not on disk after download!"
+                if check:
+
+                    response = session.get(url + ".xml")
+                    response.raise_for_status()
+                    file_metadata = self._parse_hdfxml(response)
+
+                    with open(filename_temp, "rb") as openfile:
+                        checksum = cksum(openfile)
 
-            except (HTTPError, AssertionError) as e:
-                return (url, e)
+                    assert filename_temp.stat().st_size == file_metadata["FileSize"]
+                    assert checksum == file_metadata["Checksum"]
+                # QUESTION: should we remove temp file?
+                shutil.move(filename_temp, filename)
+
+            except (HTTPError, AssertionError, FileNotFoundError) as e:
+                return (file_id, e)
```

Comment:
Remove temp file if exists?

Reply:
Don't get it ... please explain.

There's no temp file if the file exists (and overwrite is False)

Reply:
I mean: if downloading fails, there might be a partial download sitting in a temp file on disk. That temp file (with the modapedl suffix) should be deleted